src/odoc/resolver.ml

```diff
@@ -410,12 +439,60 @@ let add_unit_to_cache u =
   in
   Hashtbl.add unit_cache target_name [ u ]
 
+let lookup_path _ap ~pages ~libs:_ ~hierarchy (kind, tag, path) =
+  let module Env = Odoc_xref2.Env in
+  let ( >>= ) x f = match x with Some x' -> f x' | None -> None in
+  let page_path_to_path path =
+    (* Turn [foo/bar] into [foo/page-bar.odoc]. *)
+    let segs =
+      match List.rev path with
+      | [] -> []
+      | name :: rest -> List.rev (("page-" ^ name ^ ".odoc") :: rest)
+    in
+    Fs.File.of_segs segs
+  in
+  let find_by_path ?root named_roots path =
+    match Named_roots.find_by_path ?root named_roots ~path with
+    | Ok x -> x
+    | Error (NoPackage | NoRoot) -> None
+  in
+  let load_page path =
+    match load_unit_from_file path with
+    | Some (Odoc_file.Page_content page) -> Env.Path_page page
+    | _ -> Env.Path_not_found
+  in
+  let find_page ?root path =
+    (pages >>= fun pages -> find_by_path ?root pages path) |> function
+    | Some path -> load_page path
+    | None -> Env.Path_not_found
+  in
+  let find_page_in_hierarchy path =
+    match hierarchy with
+    | Some hierarchy -> (
+        match Hierarchy.resolve_relative hierarchy path with
+        | Ok path -> load_page path
```

Comment:
```
if Sys.is_directory path then Env.Path_directory else load_page path
```
(this check should probably be in `load_page` maybe)

Reply:
Yes!

Though this would be detected while loading the page, but the error is not propagated and is instead printed as a warning outside of the usual warning mechanisms.

Reply:
No! I think this should be reported by the `load_page` function. `Sys.is_directory` is just an other point where an exception can be raised and an other inconsistency.
That code have a TODO that I would leave for later.

Reply:
I can't see a TODO yet on this specific error...

The problem is that if you have a directory with the specific name `page-directory.odoc`, then you might have an uncaught exception when resolving `{!./page-directory}`:
```
Error while unmarshalling "h/pkg/doc/page-directory.odoc": Sys_error("Is a directory")
```

I'm happy to leave a better error reporting for later, but I would really prefer to merge without the risk of uncaught exceptions when resolving references... (note that when we lookup for assets, the right of having a directory referenced will be much higher).

Reply:
That's not an uncaught exception but a very badly reported expected error.

I agree this should be checked eventually but without a way to propagate the errors correctly, this will just slightly improve a very bad error message.
In any case, I don't want to spam the code with this kind of checks. This check must happen just before loading the file, not elsewhere.

Reply:
Oh indeed, that's not an uncaught exception. It was obvious from the message and I missed the "catch all exceptions" in `load_`...

Thanks for doing the refactoring!

Reply:
I changed the interface of the resolver completely in order to pass the errors. Which I leave for later.